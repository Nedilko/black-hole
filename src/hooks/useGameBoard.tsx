import { useMemo, useState } from 'react';
import { GameBoard, IBoardWithCells, IBoardSize } from '../game/field';

export const useGameBoard = (
  size: IBoardSize,
  holesCount: number,
  onFinish: () => void,
  onOpenCell: () => void
): [IBoardWithCells, number] => {
  const board: IBoardWithCells = useMemo(
    () =>
      GameBoard.create(
        size,
        holesCount,
        () => {
          setOpenedIndexes(board.openedCellIndexes.length);
          onOpenCell();
        },
        onFinish
      ),
    [holesCount, size]
  );

  const [openedIndexesCount, setOpenedIndexes] = useState(
    board.openedCellIndexes.length
  );

  return [board, openedIndexesCount];
};
