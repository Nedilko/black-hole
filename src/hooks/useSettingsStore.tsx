import React from 'react';
import {
  SettingsDataContext,
  SettingsDispatchContext,
} from '../store/SettingsStore';
import type { ISettings, Dispatch } from '../store/SettingsStore';

const useSettingsData = () => {
  const context = React.useContext(SettingsDataContext);
  if (context === undefined) {
    throw new Error(
      'useSettingsData must be used within a SettingsContextProvider'
    );
  }
  return context;
};

const useSettingsDispatch = () => {
  const context = React.useContext(SettingsDispatchContext);
  if (context === undefined) {
    throw new Error(
      'useSettingsDispatch must be used within a SettingsContextProvider'
    );
  }
  return context;
};

const useSettingsStore = (): [ISettings, Dispatch] => {
  return [useSettingsData(), useSettingsDispatch()];
};

export { useSettingsStore, useSettingsData, useSettingsDispatch };
