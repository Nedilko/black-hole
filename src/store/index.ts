import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import gameReducer from './gameSlice';
import fieldReducer from './fieldSlice';

const rootReducer = combineReducers({
  game: gameReducer,
  field: fieldReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
