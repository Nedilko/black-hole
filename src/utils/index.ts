import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
