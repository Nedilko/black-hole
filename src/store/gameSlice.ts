import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { FieldSize } from './fieldSlice';
import type { AppDispatch } from '.';
import { renewField, setupField } from './fieldSlice';
import { loadResults } from '../utils';

export enum DIFFUCULTY {
  EASY,
  MEDIUM,
  HARD,
  CUSTOM,
}

export interface Results {
  [DIFFUCULTY.EASY]: number;
  [DIFFUCULTY.MEDIUM]: number;
  [DIFFUCULTY.HARD]: number;
  [DIFFUCULTY.CUSTOM]: number;
}

export interface GameState {
  showGameField: boolean;
  showGameControls: boolean;
  remainingCellsCount: number;
  totalCellsCount: number;
  showTime: boolean;
  showCounter: boolean;
  showSettings: boolean;
  time: number;
  isStarted: boolean;
  isFinished: boolean;
  isWon: boolean;
  difficulty: DIFFUCULTY;
  results: Results;
}

const defaultStateValue: GameState = {
  showGameField: false,
  showGameControls: false,
  remainingCellsCount: 0,
  totalCellsCount: 0,
  showTime: false,
  showCounter: false,
  showSettings: true,
  time: 0,
  isStarted: false,
  isFinished: false,
  isWon: false,
  difficulty: DIFFUCULTY.EASY,
  results: loadResults(),
};

const gameSlice = createSlice({
  name: 'game',
  initialState: defaultStateValue,
  reducers: {
    finishGame(state) {
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
    },
    showMainMenu(state) {
      state.showGameField = false;
      state.showGameControls = false;
      state.showCounter = false;
      state.showTime = false;
      state.showSettings = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    showGameField(state) {
      state.showGameControls = false;
      state.showCounter = true;
      state.showTime = true;
      state.showSettings = false;
      state.showGameField = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    updateTime(state) {
      state.time++;
    },
    startCountdown(state) {
      state.isStarted = true;
    },
    winGame(state) {
      state.isWon = true;
      state.showGameField = false;
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
      if (
        state.results[state.difficulty] > state.time ||
        state.results[state.difficulty] === 0
      ) {
        state.results[state.difficulty] = state.time;
      }
    },
    registerDifficulty(state, action: PayloadAction<DIFFUCULTY>) {
      state.difficulty = action.payload;
    },
  },
});

export default gameSlice.reducer;

export const {
  finishGame,
  showMainMenu,
  updateTime,
  winGame,
  registerDifficulty,
  startCountdown,
} = gameSlice.actions;

export const restartGame = () => (dispatch: AppDispatch) => {
  dispatch(renewField());
  dispatch(gameSlice.actions.showGameField());
};

export const startGame =
  (size: FieldSize, holesCount: number) => (dispatch: AppDispatch) => {
    dispatch(setupField({ size, holesCount }));
    dispatch(gameSlice.actions.showGameField());
  };
