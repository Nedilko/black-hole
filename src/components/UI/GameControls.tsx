import { MdReplay, MdOutlineArrowUpward } from 'react-icons/md';
import { useDispatch } from 'react-redux';
import { gameActions } from '../../store/actions';
import { settingsActions } from '../../store/actions';

const GameControls = () => {
  const dispatch = useDispatch();

  const handleBackToMenu = () => {
    dispatch(gameActions.hideGameField());
    dispatch(gameActions.hideGameControls());
  };

  const handleTryAgain = () => {
    dispatch(gameActions.hideGameControls());
    dispatch(settingsActions.setLastSettings());
  };

  return (
    <div className="flex flex-row gap-x-8 upppercase">
      <button
        onClick={handleTryAgain}
        className="flex text-3xl items-center text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
      >
        <MdReplay />
        try again
      </button>
      <button
        onClick={handleBackToMenu}
        className="flex text-3xl items-center text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
      >
        main menu
        <MdOutlineArrowUpward />
      </button>
    </div>
  );
};

export default GameControls;
