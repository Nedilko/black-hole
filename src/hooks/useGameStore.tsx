import React from 'react';
import { GameDataContext, GameDispatchContext } from '../store/GameStore';
import type { IGameData, Dispatch } from '../store/GameStore';

const useGameData = () => {
  const context = React.useContext(GameDataContext);
  if (context === undefined) {
    throw new Error('useGameData must be used within a GameContextProvider');
  }
  return context;
};

const useGameDispatch = () => {
  const context = React.useContext(GameDispatchContext);
  if (context === undefined) {
    throw new Error(
      'useGameDispatch must be used within a GameContextProvider'
    );
  }
  return context;
};

const useGameStore = (): [IGameData, Dispatch] => {
  return [useGameData(), useGameDispatch()];
};

export { useGameStore, useGameData, useGameDispatch };
