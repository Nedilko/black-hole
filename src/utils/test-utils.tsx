import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
