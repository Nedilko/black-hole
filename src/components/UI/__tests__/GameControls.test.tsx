import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import GameControls from '../GameControls';
import { act } from 'react-dom/test-utils';
import {
  finishGame,
  GameState,
  restartGame,
  showGameField,
  startGame,
} from '../../../store/gameSlice';
import gameReducer from '../../../store/gameSlice';
import type { AppDispatch, RootState } from '../../../store/';
import { renewField } from '../../../store/fieldSlice';
import userEvent from '@testing-library/user-event';
import * as hooks from '../../../hooks';
import * as actions from '../../../store/gameSlice';

jest.mock('../../../hooks');

describe('GameControls', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render GameControls', () => {
    const { container } = renderWithProviders(<GameControls />);
    expect(container).toMatchSnapshot();
  });

  it('should render game controls buttons', () => {
    renderWithProviders(<GameControls />);
    expect(screen.getByText(/try again/)).toBeInTheDocument();
    expect(screen.getByText(/main menu/)).toBeInTheDocument();
  });

  it('should call restart game action', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());

    renderWithProviders(<GameControls />);

    const restartAction = jest.spyOn(actions, 'restartGame');

    userEvent.click(screen.getByText(/try again/));

    expect(restartAction).toHaveBeenCalledTimes(1);
  });

  it('should call show main menu action', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());

    renderWithProviders(<GameControls />);

    const showMainMenuAction = jest.spyOn(actions, 'showMainMenu');

    userEvent.click(screen.getByText(/main menu/));

    expect(showMainMenuAction).toHaveBeenCalledTimes(1);
  });

  it.skip('should dispatch restart game inner dispatches', () => {
    jest.spyOn(hooks, 'useAppDispatch').mockReturnValue(jest.fn());
    const disp = jest.fn();
    const thunk = restartGame();
    thunk(disp);
    renderWithProviders(<GameControls />);

    userEvent.click(screen.getByText(/try again/));

    expect(disp).toHaveBeenCalledTimes(2);
    expect(disp).toHaveBeenCalledWith(renewField());
    expect(disp).toHaveBeenCalledWith(showGameField());
  });
});
