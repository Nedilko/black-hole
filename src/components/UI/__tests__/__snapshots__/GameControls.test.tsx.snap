// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControls should render GameControls 1`] = `
<div>
  <div
    class="flex flex-row gap-x-8 upppercase"
  >
    <button
      class="flex text-3xl items-center text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
        />
      </svg>
      try again
    </button>
    <button
      class="flex text-3xl items-center text-gray-300/80 uppercase hover:text-shadow hover:text-gray-300 transition-all duration-150"
    >
      main menu
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
        />
      </svg>
    </button>
  </div>
</div>
`;
